java/core/src/java/org/apache/orc/impl/ColumnStatisticsImpl.java

```diff
@@ -1670,46 +1675,88 @@ public int hashCode() {
         minimum = DateUtils.convertTime(timestampStats.getMinimumUtc(),
             writerUsedProlepticGregorian, convertToProlepticGregorian, true);
       }
+      if (timestampStats.hasMaximumNanos()) {
+        maxNanos = timestampStats.getMaximumNanos();
+      }
+      if (timestampStats.hasMinimumNanos()) {
+        minNanos = timestampStats.getMinimumNanos();
+      }
     }
 
     @Override
     public void reset() {
       super.reset();
       minimum = null;
       maximum = null;
+      minNanos = null;
+      maxNanos = null;
     }
 
     @Override
     public void updateTimestamp(Timestamp value) {
       long millis = SerializationUtils.convertToUtc(TimeZone.getDefault(),
           value.getTime());
-      updateTimestamp(millis);
+      // prune the last 6 digits for ns precision
+      updateTimestamp(millis, value.getNanos() % 1_000_000);
     }
 
     @Override
-    public void updateTimestamp(long value) {
+    public void updateTimestamp(long value, int nanos) {
```

Comment:
Just curious, can we directly use java.sql.Timestamp to store minimum & maximum in UTC? If that is true, we can deprecate updateTimestamp and make the comparison and merge process easier.

Reply:
That would be useful, havent checked but I believe this could be part of a followup ticket.